go/store/datas/puller_test.go

```diff
@@ -0,0 +1,493 @@
+	require.NoError(t, err)
+
+	r, ok, err := ds.MaybeHeadRef()
+	require.NoError(t, err)
+	require.True(t, ok)
+
+	states["add big table"] = r
+
+	for k, rootRef := range states {
+		t.Run(k, func(t *testing.T) {
+			sinkdb, err := tempDirDB(ctx)
+			require.NoError(t, err)
+
+			tmpDir := filepath.Join(os.TempDir(), uuid.New().String())
+			err = os.MkdirAll(tmpDir, os.ModePerm)
+			require.NoError(t, err)
+			plr, err := NewPuller(ctx, tmpDir, 128, db, sinkdb, rootRef.TargetHash())
+			require.NoError(t, err)
+
+			err = plr.Pull(ctx)
+			require.NoError(t, err)
+
+			sinkDS, err := sinkdb.GetDataset(ctx, "ds")
+			sinkDS, err = sinkdb.FastForward(ctx, sinkDS, rootRef)
+			require.NoError(t, err)
+
+			require.NoError(t, err)
+			sinkRootRef, ok, err := sinkDS.MaybeHeadRef()
+			require.NoError(t, err)
+			require.True(t, ok)
+
+			eq, err := pullerRefEquality(ctx, rootRef, sinkRootRef, db, sinkdb)
+			require.NoError(t, err)
+			assert.True(t, eq)
+
+		})
+	}
+}
+
+func makeABigTable(ctx context.Context, db Database) (types.Map, error) {
+	m, err := types.NewMap(ctx, db)
+
+	if err != nil {
+		return types.EmptyMap, nil
+	}
+
+	me := m.Edit()
+
+	for i := 0; i < 256*1024; i++ {
+		tpl, err := types.NewTuple(db.Format(), types.UUID(uuid.New()), types.String(uuid.New().String()), types.Float(float64(i)))
+
+		if err != nil {
+			return types.EmptyMap, err
+		}
+
+		me.Set(types.Int(i), tpl)
+	}
+
+	return me.Map(ctx)
+}
+
+func pullerRefEquality(ctx context.Context, expectad, actual types.Ref, srcDB, sinkDB Database) (bool, error) {
+	expectedVal, err := expectad.TargetValue(ctx, srcDB)
+
+	if err != nil {
+		return false, err
+	}
+
+	actualVal, err := actual.TargetValue(ctx, sinkDB)
+
+	exPs, exTbls, err := parentsAndTables(expectedVal.(types.Struct))
+	actPs, actTbls, err := parentsAndTables(actualVal.(types.Struct))
+
+	if !exPs.Equals(actPs) {
+		return false, nil
+	}
+
+	err = exTbls.IterAll(ctx, func(key, exVal types.Value) error {
+		actVal, ok, err := actTbls.MaybeGet(ctx, key)
+
+		if err != nil {
+			return err
+		}
+
+		if !ok {
+			return errors.New("Missing table " + string(key.(types.String)))
+		}
+
+		exMapVal, err := exVal.(types.Ref).TargetValue(ctx, srcDB)
+
+		if err != nil {
+			return err
+		}
+
+		actMapVal, err := actVal.(types.Ref).TargetValue(ctx, sinkDB)
+
+		if err != nil {
+			return err
+		}
+
+		return errIfNotEqual(ctx, exMapVal.(types.Map), actMapVal.(types.Map))
+	})
+
+	if err != nil {
+		return false, err
+	}
+
+	return exTbls.Equals(actTbls), nil
+}
+
+var errNotEqual = errors.New("not equal")
+
+func errIfNotEqual(ctx context.Context, ex, act types.Map) error {
+	exItr, err := ex.Iterator(ctx)
+
+	if err != nil {
+		return err
+	}
+
+	actItr, err := act.Iterator(ctx)
+
+	if err != nil {
+		return err
+	}
+
+	for {
+		exK, exV, err := exItr.Next(ctx)
+
+		if err != nil {
+			return err
+		}
+
+		actK, actV, err := actItr.Next(ctx)
+
+		if err != nil {
+			return err
+		}
+
+		if actK == nil && exK == nil {
```

Comment:
This is a lot of work to say return act.Equals(exp)

Maybe those methods should be updated to handle nil receivers?